assets/src/edit-story/components/canvas/carrousel/index.js

```diff
@@ -30,12 +31,45 @@ const Page = styled.a`
 `;
 
 function Canvas() {
-	const { state: { pages, currentPageIndex }, actions: { setCurrentPageByIndex } } = useStory();
+	const { state: { pages, currentPageIndex }, actions: { setCurrentPageByIndex, arrangePage } } = useStory();
+	const getArrangeIndex = ( sourceIndex, dstIndex, position ) => {
+		// If the dropped element is before the dropzone index then we have to deduct
+		// that from the index to make up for the "lost" element in the row.
+		const indexAdjustment = sourceIndex < dstIndex ? -1 : 0;
+		if ( 'left' === position.x ) {
+			return dstIndex + indexAdjustment;
+		}
+		return dstIndex + 1 + indexAdjustment;
+	};
 	return (
 		<List>
-			{ pages.map( ( page, index ) => (
-				<Page key={ index } onClick={ () => setCurrentPageByIndex( index ) } isActive={ index === currentPageIndex } />
-			) ) }
+			{ pages.map( ( page, index ) => {
+				const onDrop = ( evt, position ) => {
+					const droppedEl = JSON.parse( evt.dataTransfer.getData( 'text' ) );
+					if ( ! droppedEl || 'page' !== droppedEl.type ) {
+						return;
+					}
+					const arrangedIndex = getArrangeIndex( droppedEl.index, index, position );
+					// Do nothing if the index didn't change.
+					if ( droppedEl.index !== arrangedIndex ) {
+						arrangePage( droppedEl.index, arrangedIndex );
+						setCurrentPageByIndex( arrangedIndex );
+					}
+				};
+				// @todo Create a Draggable component for setting data and setting "draggable".
+				const onDragStart = ( evt ) => {
+					const pageData = {
+						type: 'page',
+						index,
+					};
+					evt.dataTransfer.setData( 'text', JSON.stringify( pageData ) );
+				};
+				return (
+					<DropZone key={ index } onDrop={ onDrop } >
+						<Page draggable="true" onDragStart={ onDragStart } onClick={ () => setCurrentPageByIndex( index ) } isActive={ index === currentPageIndex } />
```

Comment:
I'm also wondering if the `draggable` a right tool here? It works great from out-of-app drag-n-drop, but it depends on the kind of UX we want to apply. E.g. we might want to highlight a place where the page will drop in the new order, shift elements a bit to demo how this element will fit, provide constraints to dragging, stuff like this. It's definitely possible to do most of it with `draggable=true`, but at that point it might seem like we doing most of it manually anyway. Maybe it'd be better to do this `Moveable`-style? We have a couple of components that do something like this: layers, page carousel, maybe animations too. So there might be savings there as well. LMK if you've already considered that approach and discarded it.

Reply:
I haven't considered using Moveable, however, I did think through the scenarios mentioned above for highlighting the area where a Page is dropped (e.g. using border for the dropzone), shifting the elements (as well doable via styling or adding an extra space element), etc. I did not find any UX related to that currently from Figma and purposefully left out handling those cases.

Would be good to confirm with @samitron7 to understand the UX for these cases.

It feels like at this moment, we can do everything we'd like to do using draggable as well, and unless there's a specific reason to create a "custom draggable" using Moveable then we could just stick to the existing option. Thoughts?

Reply:
Actually, I just saw that Figma has been updated meanwhile and there should be a blue highlight displayed indicating the place where the thumbnail will be dropped.

LMK if keeping the current approach of "draggable" seems reasonable to you as well, then I+ll continue working on the PR and implement the highlighting, too.

Reply:
Hard to say if we will run into issues with native draggable here: we might be able to solve all we need just fine. But couple of thoughts:

1. I didn't mean to use `Moveable` as is, but more to follow a similar approach: e.g. using a component with mouse/touch/pointer events + transforms. Maybe there's even an existing component we could use? Ironically I've been running into react-movable by mistake all the time, but no idea how good it is.
2. If we do use native drop events: would it be easier for us to have a single drop zone - the whole `nav` for page thumbnails and use drop coordinates to decide where the reordering will occur?
3. We also will be supporting keyboard events for this. So I'm also wondering how much we would need to do for this anyway that we'd otherwise sharable with a would-be non-native mouse-based implementation. If not a lot, then it doesn't really matter if we use pointer or drag events at this time.
4. Another thought: we have at this time three places where something like this is done: (a) page carousel, (b) page grid, and (c) layers panel. Given dnd and keyboard support, does that change at all how you'd approach this? Not 100% if all these are entirely similar, but they are at least as far as keyboard support is concerned.

Reply:
1. Hmm, I'd guess that the original element would still have to stay in place, at least in case of reordering, and we'd only need to drag the clone?
2. I did consider this, however, in this specific case, it seemed to me to be more straightforward to have a separate dropzone for each thumbnail and just checking on which side of it the element was dropped, instead of figuring out the general coordinates in the dropzone. It seems to be sufficient for any of the use cases that are currently in Figma (checking if dropped left/right/top/bottom). I suppose the issue you're thinking might be with areas which for some reason between the elements would be left without the dropzone coverage?
3. That's a good point, I'd like to understand better the UX of keyboard events for this specific case as well, e.g. if would still just the clone move and the original element stay in place. I'll look into keyboard dnd in more detail tomorrow to understand this better -- if the default draggable still makes sense.
4. I did think about the other occasions as well where we'd need dnd but not necessarily from the keyboard point of view -- I don't have experience with keyboard dnd really, TBH. Again, will look into this tomorrow to know better.

The current Dropzone/draggable implementation is definitely not final at all -- I did consider that we might need a general Draggable component (be that based on the default "draggable" or something custom then), however, I purposefully did not look into this within this PR so far to improve it as we need in the future, thinking of YAGNI while keeping in mind the use cases we have planned so far. Even if we decide to just keep the current "draggable" then it'll be very easy to replace if needed, the current implementation is not a very complex masterpiece :)

Reply:
Looked into dnd + keyboard and although I might very well not be anticipating something that you are, then it still seems to me that for now we should stick to the default "draggable".

Thinking through how the keyboard dnd would work then, in that case, we would only need to deal with keyboard events, "draggable" already solves the mouse events for us, so it does seem that there's not that much shareable between the mouse events and keyboard events logic (other than "dropping", which would already be handled in dropzones).

I will create a separate issue for keyboard page reordering once #3826 is in place and work on that separately. As for dropzones, it might still be OK to use separate dropzones for keyboard events as well, in my mind (I might be wrong) it seems to be more straightforward this way for applying for other use cases as well. WDYT?

Reply:
Okay, thought about it a bit more.

I'm thinking of reworking this to implement the originally planned DropZoneProvider which would handle `onDragOver` event to determine which dropzone within it is being hovered on and where the element is being dropped. This means that DropZoneProvider would keep track of DropZone elements within it.

This way, when doing the keyboard implementation, it could use the same logic of `onDragOver` to determine the hovered dropzone, only that this logic would be triggered by keyboard events while in dragging state (e.g. clicking left/right arrow).

Thoughts?

Reply:
Started reworking in b01dd59, still WIP.

Reply:
@miina, this sounds good. I agree - we can definitely start with draggable interface and then change if we don't like something. I did ask @pbakaus to look around if any library superficially look good to him and he found this: https://github.com/clauderic/react-sortable-hoc. It states:

> "However, most of those libraries rely on the HTML5 Drag & Drop API, which has some severe limitations. For instance, things rapidly become tricky if you need to support touch devices, if you need to lock dragging to an axis, or want to animate the nodes as they're being sorted. React Sortable HOC aims to provide a simple set of higher-order components to fill those gaps. If you're looking for a dead-simple, mobile-friendly way to add sortable functionality to your lists, then you're in the right place."

But let's keep to your plan and maybe see if any of those statement relevant to us, in case we will run into any problems in the future.